core/solver/bicg.cpp

```diff
@@ -234,12 +237,14 @@ template <typename ValueType>
 void Bicg<ValueType>::apply_impl(const LinOp *alpha, const LinOp *b,
                                  const LinOp *beta, LinOp *x) const
 {
-    auto dense_x = as<matrix::Dense<ValueType>>(x);
-
-    auto x_clone = dense_x->clone();
-    this->apply(b, x_clone.get());
-    dense_x->scale(beta);
-    dense_x->add_scaled(alpha, x_clone.get());
+    precision_dispatch_spmv<ValueType>(
```

Comment:
The solvers do not take into account the complex to real case for the simple apply. We probably need to do the same in the advanced apply?

Reply:
Do you mean when applying a real solver to a complex vector? This could be an interesting case, although I'm not sure if this is something people actually do